### PR TITLE
Custom annotations

### PR DESCRIPTION
**Pull request by donquixote**
See issue #4.

### CODE REVIEW

**Comment by slootjes** [comment]
Thanks for your effort but I'm sorry; I don't like all the shortcuts. The existing ones can be auto-completed with a good IDE and it should not become possible to mess up security rules because with these routes it's possible to have multiple while this is not possible with ``Security``. Also the RouteGET and RoutePOST do require the same amount of typing and again; adding them both will mess up the system.

I will look into the ``modifyAnnotatedRoute`` system as this looks like an improvement over the if/elseif structure there is now.

**Comment by donquixote** [comment]
> Thanks for your effort but I'm sorry; I don't like all the shortcuts.

It is up to you. If nothing else, it is a proof of concept what this interface can do.

> with these routes it's possible to have multiple while this is not possible with Security

You can have multiple `@Security` annotations on one method. Or is something currently preventing this?

> Also the RouteGET and RoutePOST do require the same amount of typing

I like these kinds of shortcuts for "find usages" or "refactor/rename", and they prevent that I misspell "POST". Well I am not really going to misspell that, but I like the feeling of using tokens known to the IDE.

If someone uses the same annotation tag twice, or two annotations that conflict with each other, we just need a clearly defined and documented behavior. E.g. should different security rules be treated as "AND" or as "OR", or should one replace the other?

I don't see a big risk of messing anything up. I rather think it is a matter of taste. Maybe a good idea is to introduce new tags or shortcuts gradually, after some time of local experimenting.
This is also what I will do. Maybe I change my mind and learn to appreciate using fewer tags with more parameters.

**Comment by slootjes** [comment]
Using multiple Security annotations will result into an exception being thrown and using shortcuts prevents this from happening without having to validate a whole bunch of extra things. Basically the same with the Method annotation where it would not be clear if would be AND or OR. The idea is to make stuff easier to do and introducing a wide range of annotations that could conflict with each other doesn't really help. I'm still considering if we could maybe introduce a ``@Content`` annotation for the (menu)titles or something if this is widely used in Drupal. 

Still I very much appreciate your input on this project, it's great to get feedback from the community on this :) Btw, I did refactor how route changes are done using a similar way you showed.

**Comment by donquixote** [comment]
> Using multiple Security annotations will result into an exception being thrown

Where does this happen? I looked into the code and tried it myself, I don't see or experience any exception being thrown.

```php
  /**
   * @Route
   * @Security(role="administrator")
   * @Security(permission="access content")
   */
  public function overview() {
```

**Comment by donquixote** [comment]
One thing we might consider is to implement some of this in the Symfony bundle instead of the Drupal module, so that more people can benefit from it.

**Comment by slootjes** [comment]
> Where does this happen? I looked into the code and tried it myself, I don't see or experience any exception being thrown.

Did you clear your caches? This is the exception that I'm getting (as expected):

```php
LogicException: Multiple "security" annotations are not allowed. in Sensio\Bundle\FrameworkExtraBundle\EventListener\ControllerListener->getConfigurations() (line 105 of vendor\sensio\framework-extra-bundle\EventListener\ControllerListener.php).
```

> One thing we might consider is to implement some of this in the Symfony bundle instead of the Drupal module, so that more people can benefit from it.

What exactly? I think the Symfony bundle is full featured as it is and some things are just not compatible so I had to redo them for Drupal.

**Comment by donquixote** [comment]
> Did you clear your caches? This is the exception that I'm getting (as expected):

I was not aware I need to visit the page to see it. I thought it would happen when clearing the cache, without necessarily visiting the page. Now I can reproduce.

This leads me to the question: Are we parsing the annotations each time someone visits the page? Is this not cached somewhere?

**Comment by donquixote** [comment]
Btw we could make those new classes implement ConfigurationInterface::getAliasName(), and let it return 'security'. This would then trigger the same exception as currently.

**Comment by slootjes** [comment]
> This leads me to the question: Are we parsing the annotations each time someone visits the page? Is this not cached somewhere?

The annotations and routes itself are cached by the annotation reader but since this system works with event listeners the annotations are read at run time. This adds some overhead obviously but it should be minimal.

> Btw we could make those new classes implement ConfigurationInterface::getAliasName(), and let it return 'security'. This would then trigger the same exception as currently.

That would actually be a valid option but still don't like the idea of having multiple annotations for the same purpose.

**Comment by donquixote** [comment]
What does Drupal currently do with multiple requirements in a route?

```
@Security(
  'access' => TRUE,
  'permission' => 'administer site configuration',
  'role' => 'administrator',
)
```

or in yml:
```
requirements:
  _permission: ..
  _role: ..
 _ access: ..
```

**Comment by slootjes** [comment]
I don't know as I actually don't work with Drupal in my daily work...if you know please explain me :) It would easily be possible to do assertions in the Security configuration to make sure just 1 of those is configured.

File: src/RouteModifier/RouteParameters.php [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+          throw new \RuntimeException("Parameter type must be a string.");
+        }
+        $parameters[$k] = ['type' => $type];
+      }
+      else {
+        throw new \RuntimeException("Parameter must be an array or a string.");
+      }
+    }
+
+    parent::__construct(['options' => $parameters]);
```
Comment (donquixote) [inline src/RouteModifier/RouteParameters.php:40]: Just saying, I think this line is wrong. Should be `['parameters' => $parameters]`.

**Comment by slootjes** [comment]
A ``@Title`` annotation is now also implemented :)

**Comment by donquixote** [comment]
Nice!

Somehow I don't like how those annotation classes are so verbose and boilerplate-y.
But I suppose it is a matter of taste, and it would be futile trying to convince you.

**Comment by slootjes** [comment]
I would rather call it explicit, a matter of taste I guess :-)

**Pull request closed without merging**